kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -303,4 +428,166 @@ private static boolean hasPhysicalName(StructField field) {
+     * @return A new {@link StructField} with updated {@link FieldMetadata}
+     */
+    private static StructField transformSchema(
+            AtomicInteger currentFieldId,
+            StructField structField,
+            String path) {
+        DataType dataType = structField.getDataType();
+        if (dataType instanceof StructType) {
+            StructType type = (StructType) dataType;
+            List<StructField> fields =
+                    type.fields().stream()
+                            .map(field -> transformSchema(
+                                    currentFieldId,
+                                    field,
+                                    getPhysicalName(field)))
+                            .collect(Collectors.toList());
+            return new StructField(
+                    structField.getName(),
+                    new StructType(fields),
+                    structField.isNullable(),
+                    structField.getMetadata());
+        } else if (dataType instanceof ArrayType) {
+            ArrayType type = (ArrayType) dataType;
+            String basePath = "".equals(path) ? getPhysicalName(structField) : path;
+            // update element type metadata and recurse into element type
+            String elementPath = basePath + "." + type.getElementField().getName();
+            structField = maybeUpdateFieldId(structField, elementPath, currentFieldId);
+            StructField elementType = transformSchema(
+                    currentFieldId, type.getElementField(), elementPath);
+            return new StructField(
+                    structField.getName(),
+                    new ArrayType(elementType),
+                    structField.isNullable(),
+                    structField.getMetadata());
+        } else if (dataType instanceof MapType) {
+            MapType type = (MapType) dataType;
+            // update key type metadata and recurse into key type
+            String basePath = "".equals(path) ? getPhysicalName(structField) : path;
+            String keyPath = basePath + "." + type.getKeyField().getName();
+            structField = maybeUpdateFieldId(structField, keyPath, currentFieldId);
+            StructField key = transformSchema(currentFieldId, type.getKeyField(), keyPath);
+            // update value type metadata and recurse into value type
+            String valuePath = basePath + "." + type.getValueField().getName();
+            structField = maybeUpdateFieldId(structField, valuePath, currentFieldId);
+            StructField value = transformSchema(currentFieldId, type.getValueField(), valuePath);
+            return new StructField(
+                    structField.getName(),
+                    new MapType(key, value),
+                    structField.isNullable(),
+                    structField.getMetadata());
+        }
+
+        return structField;
+    }
+
+    private static StructField maybeUpdateFieldId(
```

Comment:
javadocs here too, please.

Reply:
I've added some javadoc and added an example `metadata` that should make it clearer how nested ids are stored